### ISSUE TITLE
Question about the 'cross-attn per batch' part

### ISSUE THREAD

**Issue by Nightwatch-Fox11** [JiaDingCN/GeminiFusion#17]
Hello, thanks for this great work.

When debugging the code, I noticed `for bs in range(B):` in the code, so I'm wondering if there is any special reason to loop within a batch? In other words, is there any operation that prevents parallelization?

Looking forward to your reply.

**Comment by Nightwatch-Fox11** [comment]
Also, I don't understand the  '**i is in the range of d**' in the paper. According to the code, there's a loop within a batch, but I don't see anything related to '**d**'.

**Comment by gongyan1** [comment]
@Nightwatch-Fox11 You are correct. Upon careful analysis, I found that the author’s code does not align with the paper at all. The proposed pixel-wise self-attention is not reflected in the implementation. Instead, the self-attention calculated in the paper is only for the entire batch of a single sample. @ggjy @JiaDingCN 

` ## 1. 0_to_1 cross attn and skip connect
            q = x[0][bs].unsqueeze(0)  # (1,N,C)

            judger_input = torch.cat(
                [x[0][bs].unsqueeze(0), x[1][bs].unsqueeze(0)], dim=-1
            ) # (1,N,2C)

            relation_score = self.relation_judger(judger_input) # (1,N,C)

            noise_k = self.k_noise.weight[0] + q
            noise_v = self.v_noise.weight[0] + q

            k = torch.cat([noise_k, torch.mul(q, relation_score)], dim=0) # (2,N,C)
            v = torch.cat([noise_v, x[1][bs].unsqueeze(0)], dim=0) # (2,N,C)

            new_x0.append(x[0][bs] + self.cross_attn_0_to_1(q, k, v)[0].squeeze(0)) # (1,N,C)`

**Comment by gongyan1** [comment]
Upon careful analysis, I found that the author’s code does not align with the paper. The proposed pixel-wise self-attention is not reflected in the implementation. Instead, the self-attention calculated in the paper is only for the entire batch of a single sample. Have you finally resolved the issue?

**Comment by JiaDingCN** [comment]
The reason for looping within a batch is: we replace the original batch dim in the attn. The new batch dim will be the concante between tokens from different modals with the same postition. That is how we implement the pixel-to-pixel attn. It happens within a sample.

**Comment by gongyan1** [comment]
Thank you for your reply. However, I still have a doubt. Currently, the code I see is iterating over B, and in the comments after the code, I have annotated the dimensions of each variable. For self.cross_attn_0_to_1(q, k, v), the shape of the input q is (1, N, C), while the shapes of k and v are (2, N, C). Since the definition of self.cross_attn_0_to_1 is self.cross_attn_1_to_0 = nn.MultiheadAttention(dim, self.cross_heads, dropout=0.0, batch_first=False), the input received is (1, N, C). I believe the positional information mentioned in your paper should be iterated over the dimension N, but it seems that this code does not implement that.



`B, N, C = x[0].shape
Skip some codes
        new_x0 = []
        new_x1 = []
        for bs in range(B):
            q = x[0][bs].unsqueeze(0)  # (1,N,C)

            judger_input = torch.cat(
                [x[0][bs].unsqueeze(0), x[1][bs].unsqueeze(0)], dim=-1
            ) # (1,N,2C)

            relation_score = self.relation_judger(judger_input) # (1,N,C)

            noise_k = self.k_noise.weight[0] + q
            noise_v = self.v_noise.weight[0] + q

            k = torch.cat([noise_k, torch.mul(q, relation_score)], dim=0) # (2,N,C)
            v = torch.cat([noise_v, x[1][bs].unsqueeze(0)], dim=0) # (2,N,C)

            new_x0.append(x[0][bs] + self.cross_attn_0_to_1(q, k, v)[0].squeeze(0)) # (1,N,C)`

**Comment by gongyan1** [comment]
Thank you for your reply. However, I still have a doubt. Currently, the code I see is iterating over B, and in the comments after the code, I have annotated the dimensions of each variable. For self.cross_attn_0_to_1(q, k, v), the shape of the input q is (1, N, C), while the shapes of k and v are (2, N, C). Since the definition of self.cross_attn_0_to_1 is self.cross_attn_1_to_0 = nn.MultiheadAttention(dim, self.cross_heads, dropout=0.0, batch_first=False), the input received is (1, N, C). I believe the positional information mentioned in your paper should be iterated over the dimension N, but it seems that this code does not implement that.

`B, N, C = x[0].shape
Skip some codes
new_x0 = []
new_x1 = []
for bs in range(B):
        q = x[0][bs].unsqueeze(0) # (1,N,C)
        judger_input = torch.cat(
            [x[0][bs].unsqueeze(0), x[1][bs].unsqueeze(0)], dim=-1
        ) # (1,N,2C)
        relation_score = self.relation_judger(judger_input) # (1,N,C)
        noise_k = self.k_noise.weight[0] + q
        noise_v = self.v_noise.weight[0] + q
        k = torch.cat([noise_k, torch.mul(q, relation_score)], dim=0) # (2,N,C)
        v = torch.cat([noise_v, x[1][bs].unsqueeze(0)], dim=0) # (2,N,C)
        new_x0.append(x[0][bs] + self.cross_attn_0_to_1(q, k, v)[0].squeeze(0)) # (1,N,C)`

**Comment by gongyan1** [comment]
`B, N, C = x[0].shape 

Skip some codes 

new_x0 = [] 
new_x1 = [] 

for bs in range(B): 

q = x[0][bs].unsqueeze(0) # (1,N,C) 

judger_input = torch.cat( [x[0][bs].unsqueeze(0), x[1][bs].unsqueeze(0)], dim=-1 ) # (1,N,2C) 

relation_score = self.relation_judger(judger_input) # (1,N,C) 

noise_k = self.k_noise.weight[0] + q 

noise_v = self.v_noise.weight[0] + q k = torch.cat([noise_k, torch.mul(q, relation_score)], dim=0) # (2,N,C) 

v = torch.cat([noise_v, x[1][bs].unsqueeze(0)], dim=0) # (2,N,C) 

new_x0.append(x[0][bs] + self.cross_attn_0_to_1(q, k, v)[0].squeeze(0)) # (1,N,C)
`

**Comment by gongyan1** [comment]
```
        new_x0 = []
        new_x1 = []
        for bs in range(B):
            ## 1. 0_to_1 cross attn and skip connect
            q = x[0][bs].unsqueeze(0)  # (1,N,C)

            judger_input = torch.cat(
                [x[0][bs].unsqueeze(0), x[1][bs].unsqueeze(0)], dim=-1
            ) # (1,N,2C)

            relation_score = self.relation_judger(judger_input) # (1,N,C)

            noise_k = self.k_noise.weight[0] + q
            noise_v = self.v_noise.weight[0] + q

            k = torch.cat([noise_k, torch.mul(q, relation_score)], dim=0) # (2,N,C)
            v = torch.cat([noise_v, x[1][bs].unsqueeze(0)], dim=0) # (2,N,C)

            new_x0.append(x[0][bs] + self.cross_attn_0_to_1(q, k, v)[0].squeeze(0)) # (1,N,C)
```

**Comment by gongyan1** [comment]
![Image](https://github.com/user-attachments/assets/625deba7-16ce-4e91-9587-210bf456bcd6)

From this formula, we can see that the shape of each x[i] and y[i] should be N x 1, since i iterates over the range of d.

**Comment by JiaDingCN** [comment]
>  I believe the positional information mentioned in your paper should be iterated over the dimension N
 Actually, the iteration is over the dimension 0. The key idea is simple: measure the attention between two patches from different modals. So the attention is called “ per-pixel”.
